### PR TITLE
Enable python to run against other .NET repositories

### PR DESCRIPTION
**Pull request by jorive**
This is the script usage.
```log
$ py.exe .\scripts\benchmarks_ci.py --help
usage: benchmarks_ci.py [-h] [--architecture {x64,x86}]
                        [--channel {master,release/2.2.1xx,release/2.1,release/2.0.0}]
                        -f
                        [{netcoreapp3.0,netcoreapp2.2,netcoreapp2.1,netcoreapp2.0,net461} [{netcoreapp3.0,netcoreapp2.2,netcoreapp2.1,netcoreapp2.0,net461} ...]]
                        [--enable-hardware-counters]
                        [--category {coreclr,corefx}]
                        [--corerun-path CORERUN_PATH] [--cli CLI]
                        [--bdn-arguments BDN_ARGUMENTS]
                        [--optimization-level {tiered,full_opt,min_opt}]
                        [--cli-source-info {init-tools,repo,cli}]
                        [--cli-branch CLI_BRANCH]
                        [--cli-commit-sha CLI_COMMIT_SHA]
                        [--cli-repository CLI_REPOSITORY]
                        [--cli-source-timestamp CLI_SOURCE_TIMESTAMP]
                        [--generate-benchview-data]
                        [--upload-to-benchview-container UPLOAD_TO_BENCHVIEW_CONTAINER]
                        [--benchview-submission-name BENCHVIEW_SUBMISSION_NAME]
                        [--benchview-run-type {rolling,private,local}]
                        [--benchview-config-name BENCHVIEW_CONFIG_NAME]
                        [--benchview-machinepool BENCHVIEW_MACHINEPOOL]
                        [--benchview-config key value] [-q]

Tool to run .NET micro benchmarks

optional arguments:
  -h, --help            show this help message and exit
  --architecture {x64,x86}
                        Architecture of DotNet Cli binaries to be installed.
  --channel {master,release/2.2.1xx,release/2.1,release/2.0.0}
                        Download DotNet Cli from the Channel specified.
  -f [{netcoreapp3.0,netcoreapp2.2,netcoreapp2.1,netcoreapp2.0,net461} [{netcoreapp3.0,netcoreapp2.2,netcoreapp2.1,netcoreapp2.0,net461} ...]], --frameworks [{netcoreapp3.0,netcoreapp2.2,netcoreapp2.1,netcoreapp2.0,net461} [{netcoreapp3.0,netcoreapp2.2,netcoreapp2.1,netcoreapp2.0,net461} ...]]
                        Target frameworks to publish for.
  --enable-hardware-counters
                        Enables the following performance metric counters:
                        BranchMispredictions+CacheMisses+InstructionRetired
  --category {coreclr,corefx}
  --corerun-path CORERUN_PATH
                        Path to CoreRun.exe
  --cli CLI             Full path to dotnet.exe
  --bdn-arguments BDN_ARGUMENTS
                        Command line arguments to be passed to the
                        BenchmarkDotNet harness.
  --optimization-level {tiered,full_opt,min_opt}
  --cli-source-info {init-tools,repo,cli}
                        Specifies where the product information comes from.
  --cli-branch CLI_BRANCH
                        Product branch.
  --cli-commit-sha CLI_COMMIT_SHA
                        Product commit sha.
  --cli-repository CLI_REPOSITORY
                        Product repository.
  --cli-source-timestamp CLI_SOURCE_TIMESTAMP
                        Product timestamp of the soruces used to generate this
                        build (date-time from RFC 3339, Section 5.6.
                        "%Y-%m-%dT%H:%M:%SZ").
  --generate-benchview-data
                        Flags indicating whether BenchView data should be
                        generated.
  --upload-to-benchview-container UPLOAD_TO_BENCHVIEW_CONTAINER
                        Name of the Azure Storage Container to upload to.
  --benchview-submission-name BENCHVIEW_SUBMISSION_NAME
                        BenchView submission name.
  --benchview-run-type {rolling,private,local}
                        BenchView submission type.
  --benchview-config-name BENCHVIEW_CONFIG_NAME
                        BenchView's (user facing) configuration display name.
  --benchview-machinepool BENCHVIEW_MACHINEPOOL
                        A logical name that groups test results into a single
                        *machine*.
  --benchview-config key value
                        A configuration property defined as a {key:value}
                        pair. This is used to describe the benchmark results.
                        For example, some types of configurations can be:
                        optimization level (tiered, full opt, min opt),
                        configuration (debug/release), profile (on/off), etc.
  -q, --quiet           Turns off verbosity.

Additional information: This wraps all the logic of how to build/run the .NET
micro benchmarks, acquire tools, gather data into BenchView format and upload
it, archive results, etc. This is meant to be used on CI runs and available
for local runs, so developers can easily reproduce what runs in the lab. NOTE:
The micro benchmarks themselves can be built and run using the DotNet tools.
```

I have tested the scripts locally against different scenarios locally, but I am missing running them on Linux. All major scripts are can be executed stand alone, and can also be imported as modules for easy debugging/testing.
I have also run linters for correctness of the scripts, and I would recommend writing a CI around the scripts in the near future.

### CODE REVIEW

**Comment by jorive** [comment]
CoreClr command example:
```cmd
rem cd "CoreClr Repo Root"
.\build.cmd x64 release skiptests skipbuildpackages
.\run.cmd build -Project="tests\build.proj" -BuildOS=Windows_NT -BuildType=Release -BuildArch=x64 -BatchRestorePackages
.\tests\runtest.cmd x64 release generatelayoutonly
rem cd "Performance Repo Root"
py.exe .\scripts\benchmarks_ci.py --category coreclr -v -f netcoreapp3.0 --corerun-path 
"<CoreClrRepoRoot>\bin\tests\Windows_NT.x64.Release\Tests\Core_Root\CoreRun.exe" --generate-benchview-data --benchview-submission-name "This is a local run for coreclr"
```

**Comment by jorive** [comment]
@brianrob @adamsitnik @adiaaida @maririos @DrewScoggins PTAL
I finally got the chance to update all script to cover CI and dev runs. Please give them a try. The only missing part is testing on Linux.

File: scripts/dotnet.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+def add_arguments(parser: ArgumentParser) -> ArgumentParser:
+    '''
+    Adds new arguments to the specified ArgumentParser object.
+    '''
+
+    if not isinstance(parser, ArgumentParser):
+        raise TypeError('Invalid parser.')
+
+    SUPPORTED_ARCHITECTURES = [
+        'x64',  # Default architecture
```
Comment (adamsitnik) [inline scripts/dotnet.py]:
```cmd
py.exe .\scripts\benchmarks_ci.py --architecture x86 -f netcoreapp3.0 --category CoreCLR
py.exe .\scripts\benchmarks_ci.py --architecture x64 -f netcoreapp3.0 --category CoreCLR
```

Expected: runs as x86, then x64
Actual: runs twice as x86 (it was provided as first)

File: scripts/benchmarks_ci.py [hunk 2]
```diff
@@ -1,256 +1,464 @@
...
+
+    # Download DotNet Cli
+    dotnet.add_arguments(parser)
+
+    # Restore/Build/Run functionality for MicroBenchmarks.csproj
+    micro_benchmarks.add_arguments(parser)
+
+    # .NET Runtime Options.
+    parser.add_argument(
+        '--optimization-level',
```
Comment (adamsitnik) [inline scripts/benchmarks_ci.py]: tested, works. To my surprise ;p

File: scripts/micro_benchmarks.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+    parser = add_arguments(parser)
+    return parser.parse_args(args)
+
+
+def build(
+        configuration: str,
+        frameworks: list,
+        verbose: bool) -> None:
+    '''Restores and builds the benchmarks'''
+    __log_script_header("Removing packages, bin and obj folders.")
```
[the remark this answers is not in the document]
Comment (jorive) [reply scripts/micro_benchmarks.py]: On automation, it's cleaner to always start with a clean state, but I see your use case and this is an easy fix.

---
In reply to: [237081418](https://github.com/dotnet/performance/pull/150#discussion_r237081418) [](ancestors = 237081418)

File: scripts/dotnet.py (hunk 1 above)
Comment (jorive) [reply scripts/dotnet.py]: Good catch! I'll fix it!

---
In reply to: [237097443](https://github.com/dotnet/performance/pull/150#discussion_r237097443) [](ancestors = 237097443)

File: scripts/dotnet.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+    parser.add_argument(
+        '--architecture',
+        dest='architecture',
+        required=False,
+        default=SUPPORTED_ARCHITECTURES[0],
+        choices=SUPPORTED_ARCHITECTURES,
+        help='Architecture of DotNet Cli binaries to be installed.')
+
+    SUPPORTED_CHANNELS = [
+        'master',  # Default channel
```
[the remark this answers is not in the document]
Comment (jorive) [reply scripts/dotnet.py]: Is this the right mapping?
  `netcoreapp3.0` -> `master`
  `netcoreapp2.2` -> `release/2.2.1x`
  `netcoreapp2.1` -> `LTS`
  `netcoreapp2.0` -> `release/2.0`

---
In reply to: [237080255](https://github.com/dotnet/performance/pull/150#discussion_r237080255) [](ancestors = 237080255)

File: scripts/micro_benchmarks.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+        action='store_true',
+        help='''Enables the following performance metric counters:
+            BranchMispredictions+CacheMisses+InstructionRetired'''
+    )
+
+    parser.add_argument(
+        '--category',
+        required=False,
+        choices=['coreclr', 'corefx'],
+        type=str.lower)
```
[the remark this answers is not in the document]
Comment (jorive) [reply scripts/micro_benchmarks.py]: For simplicity and sanity of the script I actually trimmed the exposure/duplication of the command line options that BenchmarkDotNet has. If you feel this is a commonly used scenario, then I can add it back.

---
In reply to: [237075262](https://github.com/dotnet/performance/pull/150#discussion_r237075262) [](ancestors = 237075262)

[the remark this answers is not in the document]
Comment (jorive) [reply scripts/micro_benchmarks.py]: Optionally, users can always passed it using `--bdn-arguments`. For example:
```cmd
--bdn-arguments="--filter *Adams*"
```

---
In reply to: [237329755](https://github.com/dotnet/performance/pull/150#discussion_r237329755) [](ancestors = 237329755,237075262)

**Comment by maririos** [comment]
@jorive what is the expected usage for this script? Is it going to be consumed by all repos that want to run benchmarks?

File: scripts/dotnet.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (maririos) [reply scripts/dotnet.py]: What is a channel? Is it a branch?

**Comment by jorive** [comment]
This is a unified way of restoring/building/running/uploading/archiving? the micro benchmarks. They are targeted for automation and reproducibility of what happens in the lab. Currently all this logic is baked in the groovy files and developers need to dig for information on the CI run logs of the `perf.groovy` files. These scripts solve this problem.
I will be calling the scripts from CoreClr/CoreFx? (respective PRs will be sent after this get merged), but ultimately, it is up to the different dotnet repo owners to decide if they want/need to call the scripts.

---
In reply to: [442910908](https://github.com/dotnet/performance/pull/150#issuecomment-442910908) [](ancestors = 442910908)

File: scripts/dotnet.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jorive) [reply scripts/dotnet.py]: This is a forward parameter to the `dotnet-install.[ps1|sh]` scripts.
From [https://raw.githubusercontent.com/dotnet/cli/master/scripts/obtain/dotnet-install.ps1](https://raw.githubusercontent.com/dotnet/cli/master/scripts/obtain/dotnet-install.ps1):

```log
.PARAMETER Channel
    Default: LTS
    Download from the Channel specified. Possible values:
    - Current - most current release
    - LTS - most current supported release
    - 2-part version in a format A.B - represents a specific release
          examples: 2.0; 1.0
    - Branch name
          examples: release/2.0.0; Master
```

---
In reply to: [237574384](https://github.com/dotnet/performance/pull/150#discussion_r237574384) [](ancestors = 237574384)

File: scripts/micro_benchmarks.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+        return frameworks
+
+    @staticmethod
+    def __get_framework_channel_map() -> dict:
+        # TODO: Can we do better? For example, read it from csproj?
+        return {
+            'netcoreapp3.0': 'master',
+            'netcoreapp2.2': 'release/2.2.2xx',
+            'netcoreapp2.1': 'release/2.1',
+            'netcoreapp2.0': 'release/2.0.0',
```
Comment (adamsitnik) [inline scripts/micro_benchmarks.py]: with `release/2.0.0` it downloads `2.0.3-servicing-007156` and fails to build for netcoreapp2.0 with a really strange error:

```log
The specified framework 'Microsoft.NETCore.App', version '2.0.3' was not found.
```
the fix is to use `2.0` (I tested that it works)

```suggestion
            'netcoreapp2.0': '2.0',
```

File: src/benchmarks/micro/Program.cs [hunk 7]
```diff
@@ -28,6 +29,8 @@ private static IConfig GetConfig()
                     .WithMinIterationCount(15)
                     .WithMaxIterationCount(20) // we don't want to run more that 20 iterations
                     .AsDefault()) // tell BDN that this are our default settings
+                .WithArtifactsPath(Path.Combine(
```
[the remark this answers is not in the document]
Comment (adamsitnik) [reply src/benchmarks/micro/Program.cs]: After the offline conversation and explanation (BDN overwrites the files) I am ok with this change. #solved

**Comment by adamsitnik** [comment]
I tested in on MacOS and it works great

**Pull request merged**